### PR TITLE
Add linter rules for import order and type member delimiters

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Add linter to enforce import order and type member delimiters. I made a proposal on the rule options, but I am open to tweak them.

The important changes are on the `.eslintrc.json` file. The rest are mostly running `npm run fix` and some minor manual fixes.

Probably this PR does not need QA, since it is mainly reordering.

#### Ticket Link
None

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by enahum** [comment]
@larkox for the imports, can you make it so it is:
1. `@node_modules/alphabetical`
2. `node_modules/alphabetical`
3. `@local-alias/alphabetical`
4. `relative-path/alphabetical`
5. `./same-folder-alphabetical`

Normally this will use a line in between.

Now the best thing wouls be to have the same sort order for the import type.

Let me know if you need me to provide an example.

**Comment by larkox** [comment]
Thanks @enahum 
It should be possible. The only think I don't remember seeing is `@node_modules/` imports (I thought all of the `@` imports where aliases). That may be the trickiest part. Right now (after my latest commit) the order is the default, plus the two groups I created:
1. `builtin`
2. `external` (here would go the node_modules imports)
3. `@{**, */**}` (local aliases)
4.  `app/` (absolute path from app)
5. `parent` (relative paths going up)
6. `sibling` (relative paths in the same folder)
7. `index`

We can create any group based on minmatch patterns, but I think everything is fitting now in the order you want.

**Comment by larkox** [comment]
Also, every distinct group is ordered alphabetically internally and has empty lines separating from the other groups.

**Comment by enahum** [comment]
Yes there are like `@react-native-cookies/cookies` and many others

As for the types I think we are missing that group.

I'll post an example in about 2 or 3 hours :)

**Comment by larkox** [comment]
I made an improvement based on what you said. Sadly, there is no way to easily differentiate between aliases and packages with minmatch patterns, so this has to be done "manually". I got the list of this kind of packages from the package.json file. If we ever add a new package like that, we would have to update the `.eslintrc.json` file accordingly.

The new order would be:
1. `builtin`
2. `@node_modules` imports
2. `external` (here would go the node_modules imports)
3. `@{**, */**}` (local aliases)
4.  `app/` (absolute path from app)
5. `parent` (relative paths going up)
6. `sibling` (relative paths in the same folder)
7. `index`

File: app/actions/views/channel.js
Comment (avinashlng1080) [inline app/actions/views/channel.js:41]: We have `@contants`.   For those cases, where it has been missed, perhaps it can be replaced.

File: app/actions/views/login.js
Comment (avinashlng1080) [inline app/actions/views/login.js:19]: To replace `app/utils` with `@utils`

File: app/actions/views/select_server.test.js
Comment (avinashlng1080) [inline app/actions/views/select_server.test.js:11]: To be replaced with `@constants`

File: app/actions/views/select_team.js
Comment (avinashlng1080) [inline app/actions/views/select_team.js:16]: `@selectors` ?

File: app/actions/websocket/channels.test.js
Comment (avinashlng1080) [inline app/actions/websocket/channels.test.js:12]: I think the test helpers should also be under a directory `@test`

File: app/screens/thread/thread_base.js
Comment (avinashlng1080) [inline app/screens/thread/thread_base.js:14]: To use `@components`

**Review by avinashlng1080 — CHANGES_REQUESTED** [review]
Although I haven't covered all the files, I notice  some of the imports  that can be improved.  For example, importing from `@components` instead from `app/components` etc..

**Comment by larkox** [comment]
@avinashlng1080 I have updated all the `app/**` paths that have an alias with the aliases. It was a "replace all" approach, so I hope I haven't break anything. There are still a few paths used in the project without alias. I leave here the list, in case you want to create aliases for those too.

- `app/mattermost_bucket`
- `app/styles`
- `app/notification_preferences`
- `app/reducers`

File: app/actions/websocket/channels.test.js
Comment (larkox) [reply app/actions/websocket/channels.test.js:12]: As far as I know, there is no `@test` alias so far in the project.

**Comment by enahum** [comment]
@larkox as promised. This is an example of the post_list component

```js
import React, {ReactElement, useCallback, useEffect, useLayoutEffect, useMemo, useRef, useState} from 'react';
import {injectIntl, intlShape} from 'react-intl';
import {DeviceEventEmitter, FlatList, Platform, RefreshControl, StyleSheet, ViewToken} from 'react-native';

import {DeepLinkTypes, NavigationTypes} from '@constants';
import * as Screens from '@constants/screen';
import {useResetNativeScrollView} from '@hooks';
import {Posts} from '@mm-redux/constants';
import EventEmitter from '@mm-redux/utils/event_emitter';
import {getDateForDateLine, isCombinedUserActivityPost, isDateLine, isStartOfNewMessages} from '@mm-redux/utils/post_list';
import {badDeepLink, errorBadChannel} from '@utils/draft';
import {emptyFunction} from '@utils/general';
import {makeExtraData} from '@utils/list_view';
import {matchDeepLink, PERMALINK_GENERIC_TEAM_NAME_REDIRECT} from '@utils/url';
import telemetry, {PERF_MARKERS} from '@telemetry';

import {INITIAL_BATCH_TO_RENDER, SCROLL_POSITION_CONFIG, VIEWABILITY_CONFIG} from './post_list_config';

import type {ActionResult} from '@mm-redux/types/actions';
import type {Theme} from '@mm-redux/types/preferences';

import CombinedUserActivity from './combined_user_activity';
import DateSeparator from './date_separator';
import MoreMessagesButton from './more_messages_button';
import NewMessagesLine from './new_message_line';
import Post from './post';
```

something like `import CookieManager from `@react-native-cookies/cookies` would be above `react`

**Comment by enahum** [comment]
Based on the changes, I get that there is no need for QA review, but at least once this is "ready" to merge, we should indeed Build the PR and try it out

**Review by josephbaylon — COMMENTED** [review]
I did an initial smoke testing of the whole app on iOS and Android and did not encounter issues. I'll wait for at least one dev approval before I give my approval. Or do another round of smoke testing if there are more changes.

File: app/components/attachment_button/index.test.js
Comment (avinashlng1080) [inline app/components/attachment_button/index.test.js:7]: @larkox  Let's create an alias for the `test` directory.  Just to prevent the unintentional sorting at line 7.

File: app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser_dependencies.ts
Comment (avinashlng1080) [inline app/components/autocomplete/slash_suggestion/app_command_parser/app_command_parser_dependencies.ts:7]: @enahum  should `import type` be a separate block at the end of all other imports block?

**Review by avinashlng1080 — COMMENTED** [review]
To use aliases for constants and types.

File: app/mm-redux/reducers/entities/channel_categories.ts
Comment (larkox) [reply app/mm-redux/reducers/entities/channel_categories.ts:12]: Not really, since these are pointing to mm-redux/constants and not app/constants. We could change this to something like `@mm-redux/constants`.

File: app/screens/gallery/gallery.test.js.disabled
Comment (avinashlng1080) [inline app/screens/gallery/gallery.test.js.disabled:9]: Shouldn't the NavigationActions import be above the shallowWithIntl line ?

Correct order should be `NavigationActions`, `shallowWithIntl` and Preferences .

**Review by avinashlng1080 — APPROVED** [review]
LGTM.  

Please just verify the file `app/screens/gallery/gallery.test.js.disabled`

File: app/screens/gallery/gallery.test.js.disabled
Comment (larkox) [reply app/screens/gallery/gallery.test.js.disabled:9]: My guess is that the linter did not go through it because it is a `.disabled` extension, and not a `.js`. I can fix it manually, or we can leave it to get fixed when we remove the `.disabled`.

**Comment by josephbaylon** [comment]
@larkox Please resolve conflicts. Thanks!

**Review by enahum — APPROVED** [review]
Boom Shaka laka!!

@larkox I have an ask, can you do the same for the `Gekidou` branch?

**Comment by larkox** [comment]
Now that the rules are defined, it should be simple. I will give it a try.

**Review by josephbaylon — APPROVED** [review]
LGTM

**Comment by josephbaylon** [comment]
@larkox Looks like there are more conflicts.

**Comment by larkox** [comment]
PR updated. @josephbaylon whenever you want we can merge.

**Comment by josephbaylon** [comment]
@larkox Sorry I was late to notify you but you can go ahead and merge once the rest of the conflicts are fixed.

**Pull request merged**
